src/common/seccomp.c

```diff
@@ -0,0 +1,98 @@
+#include "postgres.h"
+#include "miscadmin.h"
+#include "common/seccomp.h"
+
+#include <unistd.h>
+
+/*
+ * Enter seccomp mode with a BPF filter that will only allow
+ * certain syscalls to proceed.
+ */
+int
+seccomp_load_rules(PgSeccompRule *syscalls, int count)
+{
+	scmp_filter_ctx ctx;
+	int rc;
+	int i;
+
+	/* By default, any syscall not in the list will crash the process */
+	if ((ctx = seccomp_init(SCMP_ACT_KILL_PROCESS)) == NULL)
+		return -1;
+
+	for (i = 0; i < count; i++) {
+		PgSeccompRule *rule = &syscalls[i];
+		rc = seccomp_rule_add(ctx, rule->psr_action, rule->psr_syscall, 0);
+		if (rc != 0)
+			return -1;
+	}
+
+	if ((rc = seccomp_load(ctx)) != 0)
+		return -1;
+
+	return 0;
+}
+
+static void
+write_str(char *s)
+{
+	/* Best effort write to stderr */
+	(void)write(STDERR_FILENO, s, strlen(s));
+}
+
+static void
+bail(int code, char *s)
+{
+	write_str(s);
+
+	/* XXX: we don't want to run any atexit callbacks */
+	_exit(code);
+}
+
+static void
+test_seccomp_sighandler(int signum, siginfo_t *info, void *cxt)
+{
+	if (signum != SIGSYS)
+		bail(1, "bad signal number\n");
+
+	/* TODO: maybe somehow extract the hardcoded syscall number */
+	if (info->_sifields._sigsys._syscall != __NR_brk)
+		bail(1, "bad syscall number\n");
+
+	bail(0, "seccomp tests have passed\n");
+}
+
+void
+test_seccomp(void)
+{
+	struct sigaction action =
+	{
+		.sa_sigaction = test_seccomp_sighandler,
+		.sa_flags = SA_SIGINFO,
+	};
+
+	PgSeccompRule syscalls[] =
+	{
+		PG_SCMP_ALLOW(exit_group),
+		PG_SCMP_ALLOW(write),
+		/* XXX: this time we'd like to receive a real SIGSYS */
+		PG_SCMP(brk, SCMP_ACT_TRAP),
+	};
+
+	/* XXX: pqsignal() is too restrictive for our purposes */
+	if (sigaction(SIGSYS, &action, NULL) != 0)
+		bail(1, "failed to install sigsys signal handler\n");
+
+	if (seccomp_load_rules(syscalls, lengthof(syscalls)) != 0)
+		bail(1, "failed to enter seccomp bpf mode\n");
+
+	/* test the good syscall */
+	write_str("write seems to work\n");
```

Comment:
It would be nice if we could test seccomp "in vivo", in the actual WAL redo process. How about installing the SIGSYS signal handler in the actual WAL redo process? You could try to print a nice error message there before killing the process. We cannot trust the output of that, but it would be really useful for debugging purposes to get a log message saying which WAL record caused the seccomp violation.

Also, we could intentionally trigger a violation as part of the WAL redo process startup, and check that it's caught correctly. Maybe I'm paranoid, but it would give me some comfort if we tested the seccomp isolation every time, and also somehow to the page server that it is working. The page server could then in turn refuse to start up if seccomp support isn't compiled in or is not working.

Reply:
> It would be nice if we could test seccomp "in vivo", in the actual WAL redo process. How about installing the SIGSYS signal handler in the actual WAL redo process?

Done. It turned out to be more hack-ish than I anticipated, but it worked. I've put down my concerns and alternative designs.

> You could try to print a nice error message there before killing the process.

That's a good idea. Unfortunately, `seccomp_syscall_resolve_num_arch()` calls `strdup()` under the hood (why???), so I had to leave the raw syscall number.